common/lib/xmodule/xmodule/x_module.py

```diff
@@ -1066,32 +1106,41 @@ class XModuleDescriptor(HTMLSnippet, ResourceTemplates, XModuleMixin):
         'name': 'display_name',
     }
 
-    # ============================= STRUCTURAL MANIPULATION ===================
-    def __init__(self, *args, **kwargs):
-        """
-        Construct a new XModuleDescriptor. The only required arguments are the
-        system, used for interaction with external resources, and the
-        definition, which specifies all the data needed to edit and display the
-        problem (but none of the associated metadata that handles recordkeeping
-        around the problem).
-
-        This allows for maximal flexibility to add to the interface while
-        preserving backwards compatibility.
+    xmodule_attributes = (
+        'course_id',
+        'displayable_items',
+        'get_display_items',
+        'get_icon_class',
+        'get_progress',
+        'get_score',
+        'handle_ajax',
+        'student_view',
+        'public_view',
+        'get_child_descriptors',
+        'xmodule_handler',
+    )
 
-        runtime: A DescriptorSystem for interacting with external resources
+    xmodule_init_complete = False
 
-        field_data: A dictionary-like object that maps field names to values
-            for those fields.
+    def init_only_for_xmodules_converted_to_xblocks(self):
+        self.__setattr__('xmodule_init_complete', True)
+        try:
+            if hasattr(self, 'init_xmodule'):
+                self.init_xmodule()
+            self.save()
+        except Exception:  # pylint: disable=broad-except
+            pass
 
-        XModuleDescriptor.__init__ takes the same arguments as xblock.core:XBlock.__init__
+    def __getattribute__(self, name):
         """
-        super(XModuleDescriptor, self).__init__(*args, **kwargs)
-        # update_version is the version which last updated this xblock v prev being the penultimate updater
-        # leaving off original_version since it complicates creation w/o any obv value yet and is computable
-        # by following previous until None
-        # definition_locator is only used by mongostores which separate definitions from blocks
-        self.previous_version = self.update_version = self.definition_locator = None
-        self.xmodule_runtime = None
+        Provides the behavior of XModuleDesciptor._xmodule() for XModules converted to XBlocks.
+        """
+        if super(XModuleDescriptorToXBlockMixin, self).__getattribute__('xmodule_init_complete') is False:
+            if name in XModuleDescriptorToXBlockMixin.xmodule_attributes:
+                # This init happened the first time an XModule attribute was accessed via the proxy on the
+                # XModuleDescriptor. So we do it here instead.
+                self.init_only_for_xmodules_converted_to_xblocks()
+        return super(XModuleDescriptorToXBlockMixin, self).__getattribute__(name)
```

Comment:
@ormsbee @cpennington @bradenmacdonald There is one last issue for which I am still trying to come up with a clean solution. For XModuleDescriptors/XModules, the XModule object is initialized (and save() is called) when an XModule property is accessed on the Descriptor in [_xmodule()](https://github.com/edx/edx-platform/blob/release-2019-04-02-10.13/common/lib/xmodule/xmodule/x_module.py#L1199-L1233). 

I am trying to keep the behavior exactly the same. Would this be the right equivalent or is there a better option?

(For reviewers, the first solution of doing it in `bind_for_student()` wasn't the right one. This is attempt 2.)

Reply:
My recollection of how I left this was that I was intending for true XBlocks to not have the `_xmodule` property. https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/x_module.py#L1885-L1895 suggests the same.

Which behavior are you trying to preserve? The extra call to `.save`?

Reply:
@cpennington Here is an overview of approaches and issues:

1. `CapaModule` had a `[init](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/capa_base.py#L237-L305)` which needs attributes available on the `ModuleSystem`. This was called in `_xmodule()` so it worked and we cannot call this code in `ProblemBlock` `init()`.
2. I tried calling it in `bind_for_student()` after the `xmodule_runtime` is set but that gets called [here](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/courseware/views/index.py#L179) in `CoursewareIndex` when trying to get the children before `self._prefetch_and_bind_section()` which causes a query to Mongo for each `ProblemBlock`. Unless we can do the caching earlier, this init has to happen later.
3. This is option 3.
4. We could delay it even further by overriding `__getattribute__` in `ProblemBlock` instead and doing the init whenever one of the attributes which are touched by `init` is called. But then we would have to do this for each XModule we convert which uses this pattern. I see at least Video and Seq use it.

As for the `save()` call I am not sure if we need it. Do we? In my earlier solutions, there were some issues with `student_view` not saving/loading the correct state that I did not investigate further because I decided to change to approach. But I am trying to keep the logic as same as possible to be on the safe side.

Reply:
Stacktrace for 2 above.

```
      File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/views/index.py", line 153, in get
    return self.render(request)
      File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/views/index.py", line 179, in render
    if self.course.has_children_at_depth(CONTENT_DEPTH):
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 519, in has_children_at_depth
    return any(child.has_children_at_depth(depth - 1) for child in self.get_children())
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 519, in <genexpr>
    return any(child.has_children_at_depth(depth - 1) for child in self.get_children())
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 519, in has_children_at_depth
    return any(child.has_children_at_depth(depth - 1) for child in self.get_children())
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 519, in <genexpr>
    return any(child.has_children_at_depth(depth - 1) for child in self.get_children())
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 517, in has_children_at_depth
    return bool(self.get_children())
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 558, in get_children
    in super(XModuleMixin, self).get_children(usage_id_filter)
      File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/mixins.py", line 407, in get_children
    if usage_id_filter is None or usage_id_filter(usage_id)
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 568, in get_child
    child = super(XModuleMixin, self).get_child(usage_id)
      File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/mixins.py", line 393, in get_child
    child_block = self.runtime.get_block(usage_id, for_parent=self)
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1904, in get_block
    return self.get_module(self.descriptor_runtime.get_block(block_id, for_parent=for_parent))
      File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/module_render.py", line 531, in inner_get_module
    will_recheck_access=True,
      File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/module_render.py", line 894, in get_module_for_descriptor_internal
    partial(LmsFieldData, student_data=student_data),
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 721, in bind_for_student
    self.init_only_for_xmodules_converted_to_xblocks()
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 376, in init_only_for_xmodules_converted_to_xblocks
    self._init_only_for_xmodules_converted_to_xblocks()
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/capa_base.py", line 131, in _init_only_for_xmodules_converted_to_xblocks
    self.lcp = self.new_lcp(self.get_state_for_lcp())
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/capa_base.py", line 201, in new_lcp
    text = self.data
      File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/fields.py", line 518, in __get__
    if field_data.has(xblock, self.name):
      File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/field_data.py", line 176, in has
    return self._field_data(block, name).has(block, name)
      File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/field_data.py", line 200, in has
    return self._source.has(block, name)
      File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/runtime.py", line 219, in has
    return self._kvs.has(self._key(block, name))
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/split_mongo/split_mongo_kvs.py", line 146, in has
    self._load_definition()
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/split_mongo/split_mongo_kvs.py", line 182, in _load_definition
    persisted_definition = self._definition.fetch()
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/split_mongo/definition_lazy_loader.py", line 31, in fetch
    definition = self.modulestore.get_definition(self.course_key, self.definition_locator.definition_id)
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/split_mongo/split.py", line 402, in get_definition
    definition = self.db_connection.get_definition(definition_guid, course_key)
      File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/split_mongo/mongo_connection.py", line 536, in get_definition
    definition = self.definitions.find_one({'_id': key})
      File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/mongodb_proxy.py", line 53, in wrapper
    return func(*args, **kwargs)
      File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/pymongo/collection.py", line 1084, in find_one
    for result in cursor.limit(-1):
      File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/pymongo/cursor.py", line 1149, in next
    if len(self.__data) or self._refresh():
      File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/pymongo/cursor.py", line 1081, in _refresh
    self.__codec_options.uuid_representation))
```

Reply:
I took a look at `_get_student_block()` too. Given that it is only called from `render()` I do not think it would be the right hook because `render()` should only be called when rendering the views and not for other cases like grading, handlers etc.

Reply:
So, one of the intentional design choices around XBlock was that they not have anything in `__init__`, because any given XBlock could be instantiated in a bunch of different contexts (some of which only have access to a subset of the available data). You're running into exactly this problem.

Going through the various things in `CapaMixin.__init__`:
1. `self.close_date` could be a `property` or a `cached_property`, rather than an attribute set during `__init__`
2. `choose_new_seed` could probably be run at when `new_lcp` is called, since I think that's the only place it's used.
3. `lcp` could be converted to a `cached_property` so that it can be generated on first use, rather than during `__init__`, but then not re-generated on future uses.
4. I'm not sure about the comment regarding `openendedresponse`. Is that still in use in the platform?

Reply:
`openendedresponse` was an input type associated with ORA1, and no longer exists in the platform.

Reply:
@cpennington Thanks! I was hoping to do that in a later round (together with other converted XBlocks) to reduce risk but I'll take a look at getting rid of the `__init__`.